burn-import/src/onnx/ir.rs

```diff
@@ -0,0 +1,322 @@
+use std::collections::HashMap;
+use strum_macros::{Display, EnumString};
+
+pub type Shape = Vec<usize>;
+
+#[derive(Debug, Clone)]
+pub struct Argument {
+    pub name: String,
+    pub arg_type: Option<ArgType>,
+}
+
+#[derive(Debug, Clone)]
+pub enum ArgType {
+    Tensor(Tensor),
+}
+
+#[derive(Debug, Clone)]
+pub struct SparseTensor(Tensor, Tensor, Shape);
+
+#[derive(Debug, Clone)]
+pub enum AttributeValue {
+    Float32(f32),
+    Int64(i64),
+    String(String),
+    Tensor(Tensor),
+    // Graph(Graph),
+    SparseTensor(SparseTensor),
+    Float32s(Vec<f32>),
+    Int64s(Vec<i64>),
+    Strings(Vec<String>),
+    Tensors(Vec<Tensor>),
+    // Graphs(Vec<Graph>),
+    SparseTensors(Vec<SparseTensor>),
+}
+pub type Attributes = HashMap<String, AttributeValue>;
+
+// https://onnx.ai/onnx/intro/concepts.html#element-type
+
+#[derive(Debug, Clone)]
+pub enum ElementType {
+    Float32,
+    Float64,
+    Int32,
+    Int64,
+    String,
+    Float16,
+    // Bfloat16,
+    // Bool,
+    // Complex128,
+    // Complex64,
+    // Int16,
+    // Int8,
+    // Uint16,
+    // Uint32,
+    // Uint64,
+    // Uint8,
+}
+
+#[derive(Debug, Clone)]
+pub struct Tensor {
+    pub name: Option<String>,
+    pub elem_type: ElementType,
+    pub shape: Shape,
+    pub data: Option<TensorData>,
+}
+
+#[derive(Debug, Clone)]
+pub enum TensorData {
+    Float16s(Vec<f16>),
+    Float32s(Vec<f32>),
+    Float64s(Vec<f64>),
+    Int32s(Vec<i32>),
+    Int64s(Vec<i64>),
+    Strings(Vec<String>),
+}
+
+#[derive(Debug, Clone)]
+pub struct Graph {
+    pub nodes: Vec<Node>,
+    pub inputs: Vec<Argument>,
+    pub outputs: Vec<Argument>,
+    pub initializers: Vec<Argument>,
+    pub old_node_names: HashMap<String, String>,
+    pub old_input_names: HashMap<String, String>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Node {
+    pub node_type: NodeType,
+    pub name: String,
+    pub inputs: Vec<Argument>,
+    pub outputs: Vec<Argument>,
+    pub initializers: Vec<Argument>,
+    pub attrs: Attributes,
+    pub is_stateful: bool,
+}
+
+// Required by topological sort
+impl PartialEq for Node {
+    fn eq(&self, other: &Self) -> bool {
+        self.name == other.name && self.node_type == other.node_type
+    }
+}
+
+// Required by topological sort
+impl Eq for Node {}
+
+// Required by topological sort
+impl core::hash::Hash for Node {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+        self.node_type.hash(state);
+        self.inputs.hash(state);
+        self.outputs.hash(state);
+    }
+}
+
+// Required by topological sort
+impl core::hash::Hash for Argument {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+    }
+}
+
+/// The list of supported node types (ONNX operators and some extra ones)
+#[derive(Debug, Hash, Eq, PartialEq, EnumString, Clone, Display)]
+pub enum NodeType {
+    // Additional Burn operators
+    Linear,
+    Gelu,
+    Conv1d,
+    Conv2d,
```

Comment:
There is a little use for it in the code. It's nice that we know some node types do not exist in ONNX spec but for the code this separation is not need. Even we decide on using this IR for exporting, the conversion will be handled by code. Lastly, it'd be difficult to check for ONNX vs BURN op type.

Reply:
Well I think it's important to avoid as much as possible using `_ => ...` in match statements. So if you only need Burn operation in one function, it's probably best to have multiple enum? Does it make sense?

Reply:
I updated the enum and mixed all in one. I explained in the doc comment that these are mixed. To iterate my point there is no distinction for the code generation if it's burn or onnx specific node. In fact, I'll make it clear in the readme, the code generation only should care about IR data model. 

As far as the current implementation using matches, I agree it's not the best style with all catch but this is a simple first version. We will refactor when we can identify a better pattern.